### PR TITLE
feat: check permissions in realtime code and frontend

### PR DESCRIPTION
**Pull request by DerMolly**
### Component/Part
frontend / commons

### Description
This PR adds permission checks in the realtime code and the frontend

### Steps

- [x] Added implementation
- [x] Added / updated tests
- [x] I read the [contribution documentation](https://github.com/hedgedoc/hedgedoc/blob/develop/CONTRIBUTING.md) and
  made sure that:
  - My commits are signed-off to accept the DCO.
  - This PR targets the correct branch: `master` for 1.x & docs, `develop` for 2.x

### CODE REVIEW

File: commons/src/utils/permissions.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
+/*
+ * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+import { NotePermissions, SpecialGroup } from './permissions.types.js'
+
```
Comment (ErikMichelson) [inline commons/src/utils/permissions.ts:7]:
```suggestion

/**
 * Checks if the given user is the owner of a note.
 *
 * @param permissions The permissions of the note to check
 * @param user The username of the user
 * @return True if the user is the owner of the note
 */
```

File: commons/src/utils/permissions.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+export const userIsOwner = (
+  permissions: NotePermissions,
+  user?: string
+): boolean => {
+  return !!user && permissions.owner === user
+}
+export const userCanEdit = (permissions: NotePermissions, user?: string) => {
```
Comment (ErikMichelson) [inline commons/src/utils/permissions.ts:14]:
```suggestion
/**
 * Checks if the given user may edit a note.
 *
 * @param permissions The permissions of the note to check
 * @param user The username of the user
 * @return True if the user has the permission to edit the note
 */
export const userCanEdit = (permissions: NotePermissions, user?: string): boolean => {
```

File: commons/src/utils/permissions.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  const isOwner = userIsOwner(permissions, user)
+  const mayWriteViaUserPermission = permissions.sharedToUsers.some(
+    (value) => value.canEdit && value.username === user
+  )
+  const mayWriteViaGroupPermission = permissions.sharedToGroups.some(
+    (value) =>
+      value.groupName === SpecialGroup.LOGGED_IN && value.canEdit && !!user
+  )
```
Comment (ErikMichelson) [inline commons/src/utils/permissions.ts:22]: This checks for `!!user` in every iteration of the loop that runs over the sharedToGroups array. You could change this to the following:

```suggestion
  const mayWriteViaGroupPermission = !!user && permissions.sharedToGroups.some(
    (value) =>
      value.groupName === SpecialGroup.LOGGED_IN && value.canEdit
  )
```

**Review by ErikMichelson — CHANGES_REQUESTED** [review]

File: commons/src/utils/permissions.ts (hunk 2 above)
Comment (ErikMichelson) [reply commons/src/utils/permissions.ts:29]: The boolean return type is still missing

**Review by ErikMichelson — CHANGES_REQUESTED** [review]

**Comment by mrdrogdrog** [comment]
Well depends. Should they see the other people cursors or not? Because that's coupled.

**Comment by DerMolly** [comment]
I think they should. Other people should not see the cursor of the read-only user, but the read-only users should see what happens in the pad.

**Review by mrdrogdrog — APPROVED** [review]

**Review by ErikMichelson — APPROVED** [review]

**Pull request merged**